pkg/reconciler/kubernetes/tektonchain/tektonchain.go

```diff
@@ -509,3 +517,51 @@ func (r *Reconciler) markUpgrade(ctx context.Context, tc *v1alpha1.TektonChain)
 	}
 	return v1alpha1.RECONCILE_AGAIN_ERR
 }
+
+func AddControllerEnv(controllerEnvs []corev1.EnvVar) mf.Transformer {
+	return func(u *unstructured.Unstructured) error {
+		if u.GetKind() != "Deployment" || len(controllerEnvs) == 0 || u.GetName() != chainsDeploymentName {
+			return nil
+		}
+
+		d := &appsv1.Deployment{}
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, d)
+		if err != nil {
+			return err
+		}
+
+		for i, c := range d.Spec.Template.Spec.Containers {
+			existingEnv := c.Env
+			if c.Name == chainsContainerName {
+				appendNewEnv := true
+				for exisitingEnvIndex, env := range existingEnv {
+					// Check for the key, if found replace it
+					for _, v := range controllerEnvs {
+						newEnv := corev1.EnvVar{
+							Name:      v.Name,
+							Value:     v.Value,
+							ValueFrom: v.ValueFrom,
+						}
+						if env.Name == newEnv.Name {
+							existingEnv[exisitingEnvIndex] = newEnv
+							appendNewEnv = false
+							break
+						}
+						// If not found append the new env
+						if appendNewEnv {
+							d.Spec.Template.Spec.Containers[i].Env = append(d.Spec.Template.Spec.Containers[i].Env, newEnv)
+						}
+					}
+				}
+			}
+		}
+
+		unstrObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(d)
+		if err != nil {
+			return err
+		}
+
+		u.SetUnstructuredContent(unstrObj)
+		return nil
+	}
+}
```

Comment:
@PuneetPunamiya I removed my earlier comments on this setion, I see more number of changes required, hence I re-arranged  the lines on this `func` as follows,
```go

func AddControllerEnv(controllerEnvs []corev1.EnvVar) mf.Transformer {
	return func(u *unstructured.Unstructured) error {
		if u.GetKind() != "Deployment" || len(controllerEnvs) == 0 || u.GetName() != chainsDeploymentName {
			return nil
		}

		d := &appsv1.Deployment{}
		err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, d)
		if err != nil {
			return err
		}

		for i, c := range d.Spec.Template.Spec.Containers {
			if c.Name != chainsContainerName {
				continue
			}

			existingEnv := c.Env
			for _, v := range controllerEnvs {
				newEnv := corev1.EnvVar{
					Name:      v.Name,
					Value:     v.Value,
					ValueFrom: v.ValueFrom,
				}
				appendNewEnv := true
				for existingEnvIndex, env := range existingEnv {
					// Check for the key, if found replace it
					if env.Name == newEnv.Name {
						existingEnv[existingEnvIndex] = newEnv
						appendNewEnv = false
						break
					}
				}
				// If not found append the new env
				if appendNewEnv {
					existingEnv = append(existingEnv, newEnv)
				}
			}

			// update the changes into the actual container
			d.Spec.Template.Spec.Containers[i].Env = existingEnv
			break
		}

		unstrObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(d)
		if err != nil {
			return err
		}

		u.SetUnstructuredContent(unstrObj)
		return nil
	}
}

```